man/hfst-split.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH HFST-SPLIT "1" "April 2026" "HFST" "User Commands"
.SH NAME
hfst-split \- =Extract transducers from archive with systematic file names
.SH SYNOPSIS
.B hfst-split
[\fI\,OPTIONS\/\fR...] [\fI\,INFILE\/\fR]
.SH DESCRIPTION
Extract transducers from archive with systematic file names
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Only print fatal erros and requested output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.TP
\fB\-\-colour\fR[=\fI\,WHEN\/\fR]
Print in colour WHEN:
.TP
\fB\-\-color\fR[=\fI\,WHEN\/\fR]
always, never, auto (default)
.SS "Input/Output options:"
.TP
\fB\-i\fR, \fB\-\-input\fR=\fI\,INFILE\/\fR
Read input transducer from INFILE
.TP
\fB\-p\fR, \fB\-\-prefix\fR=\fI\,PRE\/\fR
Use the prefix PRE in naming output files
.TP
\fB\-e\fR, \fB\-\-extension\fR=\fI\,EXT\/\fR
Use the extension EXT in naming output files
.PP
If INFILE is omitted or \-, stdin is used.
If PRE is omitted, no prefix is used.
If EXT is omitted, .hfst is used.
The extracted files are named "PRE" + N + "EXT",
where N is the number of the transducer in the archive.
.SS "An example:"
.IP
cat transducer_a transducer_b | hfst\-split \fB\-p\fR "rule" \fB\-e\fR ".tr"
.PP
This command creates files "rule1.tr" (equivalent to transducer_a)
and "rule2.tr" (equivalent to transducer_b).
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://github.com/hfst/hfst/issues>
.PP
hfst\-split home page:
<https://github.com/hfst/hfst/wiki/HfstSplit>
.br
General help using HFST software:
<https://github.com/hfst/hfst/wiki>
.SH COPYRIGHT
Copyright \(co 2017 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
# NEWS for hfst-tools and library

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

## Noteworthy changes in 3.17.1

* fixes logic error where `hfst-lexc` would treat a warning about implicit
  alphabet in XREs as a critical error even when `-Werror` is not asked

## Noteworthy changes in 3.17.0

* API breaks: HfstTokenizer, LexcCompiler, `read_in_att_format` functions
  in HfstTransducer, HfstBasicTransducer, etc. to accommodate the following
  unicode handling changes and error reporting:
* Character "tokenisers" that used to operate on bytes or UTF-8 sequences will
  now operate on [graphemes](https://www.unicode.org/reports/tr29/tr29-23.html#Grapheme_Cluster_Boundaries),
  this means that by default arcs in FSAs are expected to match combining
  grapheme clusters, e.g. a base character plus all diacritics.
  The implementation is based on ICU implementation of `u_brk`
    * in lexc string entries, graphemes are compiled into a arcs
    * in xerox regex curly braces, as well
    * in tokenise without --multichar mode, input is tokenised by graphemens
    * etc. etc. tbd
* negative weights yield warnings now unless turned off
* tools will have more gcc-style warning control: -Wall -Werror -Wno-this-error
* lexc warnings are systematically treated as errors with -Werror now
* colourised more warnings and errors

## Noteworthy changes in 3.16.2

* Support for `>OpenFst-1.8.0`
    * only functional change is removal of n-best paths search for log weight
      automata, it probably has errored out before but has slightly different
      error now (HfstException instead of openfst exit)
* Optimisations to hfst-pmatch2fst alphabet handling

## Noteworthy changes in 3.16.1

* Bug fixes:

    * hfst-tokenize escapes both \r and \n now

    * regexp parsing more robust with comments and linebreaks

* Removal of some bundled libraries including foma and openfst, install
  suitable versions to compile support for them

* ICU stabilised as unicode handler over all

* experimental C API for rust bindings (unstable)

* new tools (testing):

    * hfst-check-alpha for checking and comparing alphabets in automata and
      their metadata

    * hfst-dump-alphabets for reviewing alphabets easier than hfst-summarize
      grepping

    * hfst-kill-paths for efficient path removal

* introduced clang-format to work towards standard code formatting

Noteworthy changes in 3.16.0
----------------------------

* Bugfixes:

  * During conversion between optimized-lookup and other formats, the alphabet
    was not being preserved, leading to incorrect behaviour in some cases.
    This has been fixed.

  * Fix segfaults in hfst-summarize, hfst-disjunct and foma back-end.

  * Force ICU, fixes #501.

  * Use unique_ptr for input/ouput stream memory allocation (#547).


* Improvements:

  * hfst-pmatch archives now no longer include or parse redundant alphabets,
    which in some cases can be a meaningful size and speed saving.

  * Multiple memory leaks fixed.

  * Drop support for TR1 standard assuming that C++14 is minimal requirement.

  * Drop legacy wrappers for foma and lexc (#532).

  * Use upstream Foma and by default upstream OpenFST.

  * Use C++17, or C++14 with bundled OpenFST.

  * Update external links for sfst, foma and flex.

  * Add lex/yacc sources to dist (see issue #524).

* Features:

  * A new variable in the hfst-pmatch language, "unicode-character-classes",
    which causes the use of built-in character sets to be interpreted at
    runtime using a unicode library.

  * Verbose output from hfst-pmatch2fst has been much improved, providing eg.
    a call graph of the process. It also warns about use of pmatch features
    that is likely to be incorrect and lead to errors.


Noteworthy changes in 3.15.5
----------------------------

* Fixes to pmatch:

  * Modify tracking of entry points to allow for RTNs in Captures()

  * Fix off-by-one bug in checking captures

* Fixes to other issues: #508, #488, #486, #452, #491, #494

* Fix memory leaks


Noteworthy changes in 3.15.4
----------------------------

* Import Foma patches with license change to Apache-2.0 and minimal other fixes

* Allow regeneration of Foma files so the generated files can be omitted

* Explicit license file for SFST


Noteworthy changes in 3.15.3
----------------------------

* Numerous improvements to command line tools, e.g. substitute, insert-freely, proc, fst2txt, expand


Noteworthy changes in 3.15.2
----------------------------

* Fix off-by-one bug in checking captures in pmatch


Noteworthy changes in 3.15.1
----------------------------

* Fix issue #443

* Fix issue #429

* Fix issue #409


Noteworthy changes in 3.15.0
----------------------------

* Numerous improvements in pmatch processing:

  * hfst-pmatch2fst now implements eg @include"filename.txt" as a preprocessing stage
  
  * hfst-pmatch2fst now implements Explode() and Implode() in full generality
  
  * hfst-pmatch now allows printing weights (in location mode, with --locate)
  
  * hfst.PmatchContainer.locate offered via python bindings
  
  * bugfixes and improvements hfst.PmatchContainer.tokenize

* Add tool hfst-eliminate-flags.

* Allow both escaped and unescaped zeros in multicharacters in all lexc lexicon entries (fixes issue #274).

* Also reweight state that is both initial and final in hfst-reweight (fixes issue #368).

* Do not let distributed files depend on configuration. Include man files in dist.

* Print warnings and error messages also in earlier stages of twol compilation.

* Add option --execute-and-quit (-E) to hfst-xfst.


Noteworthy changes in 3.14.0
----------------------------

* Numerous improvements to pmatching and tokenization:

  * pmatch now supports the default symbol

  * pmatch now supports reading word embeddings in the binary format


tools/src/hfst-split.cc
//! @file hfst-split.cc
//!
//! @brief Transducer archive exploding tool
//!
//! @author HFST Team


//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#ifdef WINDOWS
#include <io.h>
#endif

#include <iostream>
#include <fstream>
#include <memory>

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>

#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "HfstTransducer.h"
#include "HfstInputStream.h"
#include "HfstOutputStream.h"

#include "inc/globals-common.h"
#include "inc/globals-unary.h"

using hfst::HfstTransducer;
using hfst::HfstInputStream;
using hfst::HfstOutputStream;

using std::unique_ptr;
using std::make_unique;

// add tools-specific variables here
char *prefix;
char *extension;

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out, "Usage: %s [OPTIONS...] [INFILE]\n"
           "Extract transducers from archive with systematic file names\n"
        "\n", program_name);
    print_common_program_options(message_out);
    fprintf(message_out, "Input/Output options:\n"
            "  -i, --input=INFILE    Read input transducer from INFILE\n"
            "  -p, --prefix=PRE      Use the prefix PRE in "
            "naming output files\n"
            "  -e, --extension=EXT   Use the extension EXT in "
            "naming output files\n");
    fprintf(message_out, "\n");
    fprintf(message_out,
        "If INFILE is omitted or -, stdin is used.\n"
            "If PRE is omitted, no prefix is used.\n"
        "If EXT is omitted, .hfst is used.\n"
        "The extracted files are named \"PRE\" + N + \"EXT\",\n"
        "where N is the number of the transducer in the archive.\n\n"
        "An example:\n"
        "   cat transducer_a transducer_b | hfst-split -p \"rule\" -e \".tr\"\n\n"
        "This command creates files \"rule1.tr\" (equivalent to transducer_a)\n"
        "and \"rule2.tr\" (equivalent to transducer_b). \n");
    fprintf(message_out, "\n");
    print_report_bugs();
    fprintf(message_out, "\n");
    print_more_info();
}


int
parse_options(int argc, char** argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    extension = hfst_strdup(".hfst");
    prefix = hfst_strdup("");
    while (true)
    {
        static const struct option long_options[] =
        {
          HFST_GETOPT_COMMON_LONG,
          {"input", required_argument, 0, 'i'},
          {"prefix", required_argument, 0, 'p'},
          {"extension", required_argument, 0, 'e'},
          // add tool-specific options here
            {0,0,0,0}
        };
        int option_index = 0;
        // add tool-specific options here
        int c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT "i:p:e:",
                             long_options, &option_index);
        if (-1 == c)
        {
            break;
        }

        switch (c)
        {
#include "inc/getopt-cases-common.h"
        case 'i':
          inputfilename = hfst_strdup(optarg);
          inputfile = hfst_fopen(inputfilename, "r");
          if (inputfile == stdin)
            {
              free(inputfilename);
              inputfilename = hfst_strdup("<stdin>");
            }
          inputNamed = true;
          break;
        case 'p':
          free(prefix);
          prefix = hfst_strdup(optarg);
          break;
        case 'e':
          free(extension);
          extension = hfst_strdup(optarg);
          break;
#include "inc/getopt-cases-error.h"
        }
    }

#include "inc/check-params-common.h"
#include "inc/check-params-unary.h"
    return EXIT_CONTINUE;
}

int
process_stream(HfstInputStream& instream)
{
  //instream.open();
    
    size_t transducer_n=0;
    while(instream.is_good())
    {
        transducer_n++;
        outfilename = static_cast<char*>(hfst_malloc(sizeof(char) *
                             strlen(prefix) + strlen(extension) +
                             strlen("123456789012345678901234567890")));
        sprintf(outfilename, "%s" SIZE_T_SPECIFIER "%s", prefix, transducer_n,
                              extension);
        verbose_printf("Writing " SIZE_T_SPECIFIER " of %s to %s...\n", transducer_n,
                       inputfilename, outfilename);
        auto outstream = make_unique<HfstOutputStream>(outfilename, instream.get_type());
        //outstream->open();
        HfstTransducer trans(instream);
        *outstream << trans;
        outstream->flush();
        outstream->close();
        free(outfilename);
    }
    instream.close();
    return EXIT_SUCCESS;
}


int main( int argc, char **argv ) {
#ifdef WINDOWS
  _setmode(0, _O_BINARY);
#endif

    hfst_set_program_name(argv[0], "0.1", "HfstSplit");
    int retval = parse_options(argc, argv);
    if (retval != EXIT_CONTINUE)
    {
        return retval;
    }
    // close buffers, we use streams
    if (inputfile != stdin)
    {
        fclose(inputfile);
    }
    if (outfile != stdout)
    {
        fclose(outfile);
    }
    verbose_printf("Reading from %s, writing to %s...%s\n",
        inputfilename, prefix, extension);
    // here starts the buffer handling part
    unique_ptr<HfstInputStream> instream;
    try {
      instream.reset((inputfile != stdin) ?
        new HfstInputStream(inputfilename) : new HfstInputStream());
    } catch(const HfstException e)  {
        error(EXIT_FAILURE, 0, "%s is not a valid transducer file",
              inputfilename);
        return EXIT_FAILURE;
    }
    
    retval = process_stream(*instream);
    free(inputfilename);
    return retval;
}

